Tests for fallback boxed dispatch (including TLS mode)

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26719 Tests for fallback boxed dispatch (including TLS mode)**

This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_local_tensor_type_set()` which is considered as part of dispatch. 
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

ezyang committed:
Tests for fallback boxed dispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
namedtensor macrofix on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Include the correct named tensor header on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch (including TLS mode)"


This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_included_tensor_type_set()` which is considered as part of dispatch. (It is a little goofy that we now have to do two TLS accesses; I'm not too sure if there's a way to avoid this.)
* There is a screed about how modes (Profiling/Tracing) are different from hybrid wrapper-modes (Variable)
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

zdevito commented on c10/core/impl/LocalTensorTypeSet.cpp:
```diff
@@ -14,29 +14,55 @@ namespace {
 
 #endif
 
+void set_enabled_via_excluded(TensorTypeId tid, bool enabled) {
```
The names of these functions are confusing to me. The caller still needs to understand that thefinal set will be `(tensor_properties | included) - excluded` but then there is also a second translation between how this function affects included/excluded. Wouldn't it make more sense to just have it follow the set operations directely, excluded.add, excluded.remove, included.add, included.remove.

zdevito commented on c10/core/impl/LocalTensorTypeSet.h:
```diff
@@ -15,8 +15,74 @@
 C10_API void tls_variable_set_enabled(bool enabled);
-C10_API TensorTypeSet tls_excluded_tensor_type_set();
+
+C10_API bool TESTING_ONLY_tls_generic_mode_is_enabled();
```
Is it the best idea to have separate functions here for every key? I am concerned that over time non-simple operations on the tensor set can end up here, but I'd prefer the opposite: that the only thing you can do to a key is to enable it in a thread or in tensors properties.

ezyang replied:
Solves my giant comment problem too.

zdevito commented on c10/core/impl/LocalTensorTypeSet.cpp:
```diff
@@ -14,29 +14,55 @@ namespace {
 
 // NB: Zero initialized!
 thread_local uint64_t raw_excluded;
+thread_local uint64_t raw_included;
```
These are used adjacent to each other in dispatch, so it would be more efficient if they were in a single thread_local struct. Otherwise there is a sequence of like 4 loads to get to each one. I would imagine this struct would grow into the generic thread-local dispatch state that the chain of responsibility would need to call `super` as well.

zdevito reviewed (approved):
Generally looks good. I have a few medium level comments about the API and use of thread locals.

ezyang committed:
Complete rework of API on "Tests for fallback boxed dispatch (including TLS mode)"


This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_included_tensor_type_set()` which is considered as part of dispatch. (It is a little goofy that we now have to do two TLS accesses; I'm not too sure if there's a way to avoid this.)
* There is a screed about how modes (Profiling/Tracing) are different from hybrid wrapper-modes (Variable)
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Stop destructing guards too early on "Tests for fallback boxed dispatch (including TLS mode)"


This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_local_tensor_type_set()` which is considered as part of dispatch. 
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch (including TLS mode)"


This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_local_tensor_type_set()` which is considered as part of dispatch. 
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]

ezyang committed:
Update on "Tests for fallback boxed dispatch (including TLS mode)"


This PR adds a pair of tests for fallback boxed dispatch, exercising two different ways you might use it: (1) to implement a "wrapper" tensor type (e.g., LazyTensor, NestedTensor), and (2) to implement a toggleable "mode" (e.g., Profiling, Tracing). Both implement the most trivial possible implementations of their type: they "wrap" a real tensor simply forward along to the real implementation. This PR also adds the necessary feature support for toggleable mode, which is in the original generic dispatch abstraction design, but was not previously implemented. I had not originally intended to add this, but it turns out writing a new "mode" is a lot simpler than writing a "wrapper" type, so I ended up writing the mode version first.

General structure of the PR:

* Add two new testing tensor type ids, `TESTING_ONLY_GenericWrapperTensorId` and `TESTING_ONLY_GenericModeTensorId`, which our tests use. They might find other use in other tests if necessary.
* Add support for toggling the availability of `TESTING_ONLY_GenericModeTensorId`. Introduces a new thread local variable accessible by `tls_local_tensor_type_set()` which is considered as part of dispatch. 
* The mode fallback is very simple: it increments a counter and then passes on the call to the underlying kernel by invoking the JIT.
* The wrapper fallback is more complex: it parses the arguments, unwrapping any wrapped tensor arguments, then invokes the JIT, and then rewraps the outputs.

The examples here are somewhat simplistic; there are a number of engineering improvements that could be applied. We could save these for later (landing this patch to get immediate testing), or incorporate them into this patch:

* `getOperator` is horrible. Bram Wasti and I discussed a plan for how to make this easier, by simply refactoring the JIT interface.
* `GenericWrapperTensorImpl` doesn't populate all of its fields accurately. Most notably, size is not setup correctly.
* `generic_wrapper_fallback` should handle tensor lists in arguments and returns properly.

One pitfall: fallback dispatch only works with non-c10 code. That's why I test using `batch_norm`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17549624](https://our.internmc.facebook.com/intern/diff/D17549624)

[ghstack-poisoned]
